File: lib/lambdas/lambda-save-state.py
import json
import boto3
import os
import logging

logger = logging.getLogger()
logger.setLevel(logging.INFO)

ec2 = boto3.client('ec2')
ssm = boto3.client('ssm')

def handler(event, context):
	logger.info("Save state lambda triggered")
	logger.info('Event: ' + json.dumps(event))

	instance_id = event['detail']['instance-id']
	logger.info(f"Handling stop event for instance: {instance_id}")

	spotfleet_id = os.environ['CLOUDRIG_SPOTFLEET_REQUEST_ID']

	logger.info(f"Listing the instances managed by this stack spotfleet {spotfleet_id}...")
	spotfleet_instances = ec2.describe_spot_fleet_instances(SpotFleetRequestId=spotfleet_id)

	#logger.info(f"Checking if the instance {instance_id} is in this list...")
	#is_instance_in_this_stack = False
	#for instance in spotfleet_instances['ActiveInstances']:
	#	if instance['InstanceId'] == instance_id:
	#		is_instance_in_this_stack = True
	is_instance_in_this_stack = True

	if is_instance_in_this_stack:
		logger.info("Triggering the state save automation...")
		ssm.start_automation_execution(DocumentName=os.environ['CLOUDRIG_SAVE_STATE_AUTOMATION_DOCUMENT_NAME'], Parameters={'InstanceId': [instance_id]})
		logger.info('Done starting the save state automation')
	else:
		logger.info(f'Instance {instance_id} is not part of this stack. Ignoring...')

File: lib/lambdas/lambda-save-state.js
var AWS = require('aws-sdk');
var waterfall = require('async-waterfall');

exports.handler = (event, context, callback) => {

	console.log("Save state lambda triggered");
	console.log('Event: ' + JSON.stringify(event));
	console.log('Context: ' + JSON.stringify(context));

	const ec2 = new AWS.EC2();
	const ssm = new AWS.SSM();

	const instanceId = event.detail["instance-id"];
	console.log("Handling stop event for instance: " + instanceId);

	waterfall([
			function (cb) {
				console.log("Listing the instances managed by this stack spotfleet " + process.env.CLOUDRIG_SPOTFLEET_REQUEST_ID + "...");
				ec2.describeSpotFleetInstances({
					SpotFleetRequestId: process.env.CLOUDRIG_SPOTFLEET_REQUEST_ID,
				}, cb)
			},

			function (data, cb) {
				console.log("Checking if the instance " + instanceId + " is in this list...");
				var isInstanceInThisStack = false;
				for (const instance of data['ActiveInstances']) {
					if (instance['InstanceId'] === instanceId) {
						isInstanceInThisStack = true;
					}
				}

				if (isInstanceInThisStack) {
					cb();
				} else {
					console.log('Instance ' + instanceId + ' is not part of this stack. Ignoring...');
				}
			},

			function (data, cb) {
				console.log("Triggering the state save automation...");
				ssm.startAutomationExecution({
					'DocumentName': process.env.CLOUDRIG_SAVE_STATE_AUTOMATION_DOCUMENT_NAME,
					'Parameters': {
						'InstanceId': [instanceId],
					}
				}, cb);
			}
		],
		// CATCH ERRORS
		function(error, data) {
			if (error) {
				console.log('Error while executing: ' + error); callback(error);
			} else {
				console.log('Done starting the save state automation');
				callback(null, data);
			}
		})
};
